src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -24,12 +26,344 @@
+ * Finding a cycle  indicates a distributed deadlock and the function returns
+ * true on that case. Also, the deadlockPath is filled with the transaction
+ * nodes that form the cycle.
+ */
+static bool
+CheckDeadlockForTransactionNode(TransactionNode *startingTransactionNode,
+								List **deadlockPath)
+{
+	List *waitingTransactionNodes = list_copy(startingTransactionNode->waitsFor);
+
+	/* traverse the graph */
+	while (waitingTransactionNodes != NIL)
+	{
+		TransactionNode *waitingTransactionNode =
+			(TransactionNode *) linitial(waitingTransactionNodes);
+		ListCell *currentWaitForCell = NULL;
+
+		waitingTransactionNodes = list_delete_first(waitingTransactionNodes);
+
+		/* cycle found, let the caller know about the cycle */
+		if (waitingTransactionNode == startingTransactionNode)
+		{
+			*deadlockPath = list_make1(waitingTransactionNode);
+
+			FindDeadlockPath(waitingTransactionNodes, deadlockPath);
+
+			return true;
+		}
+
+		/* don't need to revisit the node again */
+		if (waitingTransactionNode->transactionVisited)
+		{
+			continue;
+		}
+
+		waitingTransactionNode->transactionVisited = true;
+
+		/* prepend to the list to continue depth-first search */
+		foreach(currentWaitForCell, waitingTransactionNode->waitsFor)
+		{
+			TransactionNode *waitForTransaction = (TransactionNode *) lfirst(
+				currentWaitForCell);
+
+			/*
+			 * We add both the waiting node and its parent to simplify finding the
+			 * deadlock path.
+			 */
+			waitingTransactionNodes =
+				lcons(waitingTransactionNode, waitingTransactionNodes);
+			waitingTransactionNodes = lcons(waitForTransaction, waitingTransactionNodes);
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * FindDeadlockPath gets a waitingTransactionList which involves a deadlock
+ * (i.e., cycle in the graph) and fills the deadlockPath by backtracing on
+ * the waiting list.
+ *
+ * The function iterates over the waiting list, adds the element to the
+ * deadlock path and removes all other branches that starts with the
+ * given node on the waiting graph.
+ */
+static void
+FindDeadlockPath(List *waitingTransactionList, List **deadlockPath)
+{
+	while (waitingTransactionList != NIL)
+	{
+		TransactionNode *waitingTransaction = linitial(waitingTransactionList);
+
+		*deadlockPath = lappend(*deadlockPath, waitingTransaction);
+
+		waitingTransactionList = list_delete_first(waitingTransactionList);
+
+		/* delete all the other paths up to waitingTransaction from the graph */
+		waitingTransactionList =
+			RemoveElementsUptoNode(waitingTransactionList, waitingTransaction);
+	}
+}
+
+
+/*
+ * RemoveElementsUptoNode gets a transaction list and a transactionNode.
+ * The function searches for the given transactionNode in the list. If it
+ * finds any, the elements up to the last found node (including it) are
+ * removed from the list. If not any found, the list returned as it is.
+ */
+static List *
+RemoveElementsUptoNode(List *waitingTransactionList,
+					   TransactionNode *transactionNode)
+{
```

Comment:
This seems quite complicated. I don't quite get why you're not emulating what the callstack previously did in the recursive version?  You can have a very similar logic to maintain the list of nodes leading to the cycle?  A bit simplified: Constantly keep track of the "depth" of the search, by putting that into the thing you push onto the 'please-visit-me' list. In parallel keep an array of 'MaxBackends' size. Whenever visiting the next node, fill the current depth with the to-be-visited node.  And voila, you should end up with the firsth depth entries of that array describing the cycle?